Refactor partition manager api

jhorstmann opened the pull request.

lukasniemeier-zalando commented on src/main/java/org/zalando/fahrschein/PartitionManager.java:
```diff
@@ -1,53 +1,28 @@
      * @param lockedBy Name of one node of the consuming application
      * @param timeout Timeout after that the lock will automatically be released
      * @param timeoutUnit Time unit for the timeout
      * @return Whether locking succeeded, i.e. the lock was not held by anyone or the existing lock expired. The current implementations will also return true if the lock is already held by the same node, but not update the expiry time.
```
Update javadoc?

lukasniemeier-zalando commented on src/main/java/org/zalando/fahrschein/PartitionManager.java:
```diff
@@ -1,53 +1,28 @@
-                result.add(partition);
-            }
-        }
+    boolean unlockPartitions(Lock lock);
```
adapt javadoc?

lukasniemeier-zalando commented:
- [ ] document usage of `PartitionManager`

tmuehl commented:
Looks good 👍

tmuehl commented on src/main/java/org/zalando/fahrschein/PartitionManager.java:
```diff
@@ -9,17 +9,19 @@
      * @param timeout Timeout after that the lock will automatically be released
      * @param timeoutUnit Time unit for the timeout
-     * @return Whether locking succeeded, i.e. the lock was not held by anyone or the existing lock expired. The current implementations will also return true if the lock is already held by the same node, but not update the expiry time.
+     * @return A {@code Lock} instance containing the list of locked partitions if locking succeded, {@code Optional.empty()} otherwise
```
If you already started with Javadoc, why not `{@link org.zalando.fahrschein.domain.Lock}`?

(you have a type: "succeeded")

tmuehl commented:
👍

tmuehl commented on src/main/java/org/zalando/fahrschein/NakadiClient.java:
```diff
@@ -37,13 +38,19 @@
         this.objectMapper = objectMapper;
         this.cursorManager = cursorManager;
-        this.nakadiReaderFactory = new NakadiReaderFactory(clientHttpRequestFactory, backoffStrategy, cursorManager, objectMapper);
+        this.nakadiReaderFactory = new NakadiReaderFactory(clientHttpRequestFactory, backoffStrategy, cursorManager, objectMapper, metricsCollector);
```
I explicitly designed the `MetricsCollector` to be passed into the `listen` method. I know this is ugly because you have so many parameters there.

But I wanted to be able to collect different metrics for different eventtypes.

jhorstmann replied:
> I wanted to be able to collect different metrics for different eventtypes.

Ok, I have to look at this again. Would an alternative be to have one MetricsCollector but change the interface so the event name is passed to all methods?

tmuehl replied:
Sounds good! Another alternative might be to have yet another factory, a `MetricsCollectorFactory`.

tmuehl commented on src/main/java/org/zalando/fahrschein/NakadiClient.java:
```diff
@@ -80,33 +86,30 @@ public Subscription subscribe(String applicationName, String eventName, String c
+
+        if (lockTimeout > 0) {
+            final Thread currentThread = Thread.currentThread();
+            executor.schedule(currentThread::interrupt, lockTimeout, timeoutUnit);
```
Nice approach!

jhorstmann replied:
> Out of curiosity, why don't you use stream_timeout?

I want to have control of the timeout on client side. Using the same value for stream and lock timeout makes sense though.

tmuehl commented on README.md:
````diff
@@ -113,7 +113,30 @@ final ManagedCursorManager cursorManager = new ManagedCursorManager(baseUri, req
+Partitions are locked by one node for a certain time. This requires that every node has an unique name or other identifier. The `listen` method will return after the lock timeout and can try to grab the lock again.
+
+```java
+@Scheduled(fixedDelay = 60*1000L)
````
Why do you delay this method?

jhorstmann replied:
> Why do you delay this method?

Good point, there are now two cases where this method can exit:

 - after the specified timeout
 - when an unchecked exception is thrown

For the 2nd case a slight delay might make sense. Another way to implement it would be with the `Async` annotation and having another loop inside the method.

tmuehl replied:
Ok, I see, makes sense.

tmuehl commented on src/main/java/org/zalando/fahrschein/NakadiReader.java:
```diff
@@ -315,7 +303,11 @@ private void expectField(JsonParser jsonParser, String expectedFieldName) throws
         final JsonToken token = jsonParser.nextToken();
         if (token == null) {
-            throw new EOFException(Thread.currentThread().isInterrupted() ? "Thread was interrupted" : "Stream was closed");
+            if (Thread.currentThread().isInterrupted()) {
```
Doesn't it need to be

```
if (Thread.currentThread().isInterrupted()) {
    throw new InterruptedIOException("Thread was interrupted");
} else {
    if (token == null) {
        throw new EOFException("Stream was closed");
}
```
?
Otherwise the stream won't stop until you also close the connection.

jhorstmann replied:
When the input stream is interrupted its read method will return -1, which will result in a `null` token. The different exception here is just to get nicer messages in logs, the `IOException` will be caught in the main loop and there is another check if the thread was interrupted.

tmuehl replied:
Ah ok, thanks for the explanation. I though the `JsonInput` would block and hence `currentThread().interrupt()` would only set the interrupted flag without killing any connection.

lukasniemeier-zalando commented on src/main/java/org/zalando/fahrschein/NakadiReader.java:
```diff
@@ -277,9 +265,9 @@ void runInternal(long timeout, TimeUnit timeoutUnit) throws IOException, Backoff
 
                 if (errorCount > 0) {
-                    LOG.warn("Got [{}] while reading events for [{}] after [{}] retries", e.getClass().getSimpleName(), eventName, errorCount, e);
+                    LOG.warn("Got [{}] [{}]while reading events for [{}] after [{}] retries", e.getClass().getSimpleName(), e.getMessage(), eventName, errorCount, e);
```
missing whitespace

jhorstmann commented:
@lukasniemeier-zalando I simplified the api a bit and added a `SimplePartitionManager` which could be used for an implementation based on global resource locks.
